File: mobile/o_o/lib/features/mindmap/domain/entities/mindmap_node.dart
import 'package:flutter/material.dart';

/// 노드 콘텐츠 타입
enum NodeContentType {
  text,
  video,  // youtube → video로 변경 (API 스펙에 맞춤)
  image,
}

/// 마인드맵 노드
///
/// 마인드맵의 각 노드를 나타냅니다.
class MindmapNode {
  /// 노드 ID (API id)
  final String id;

  /// 서버 숫자 ID (API nodeId)
  final int? nodeId;

  /// 워크스페이스 ID (API workspaceId)
  final int? workspaceId;

  /// 부모 노드 ID (API parentId)
  final int? parentId;

  /// 노드 텍스트 (API keyword)
  final String text;

  /// 노드 위치 (노드 중심의 캔버스 좌표)
  final Offset position;

  /// 노드 배경 색상 (API color)
  final Color color;

  /// 노드 너비
  final double width;

  /// 노드 높이
  final double height;

  /// 노드 레벨 (0: 중심, 1: 1차 자식, 2: 2차 자식...)
  final int level;

  /// 노드 콘텐츠 타입 (API type)
  final NodeContentType contentType;

  /// 콘텐츠 URL (YouTube URL 또는 이미지 URL - keyword에서 추출)
  final String? contentUrl;

  /// 콘텐츠 설명/메모 (API memo)
  final String? description;

  /// 분석 상태 (API analysisStatus)
  final String? analysisStatus;

  /// 생성일시 (API createdAt)
  final DateTime? createdAt;

  /// 수정일시 (API updatedAt)
  final DateTime? updatedAt;

  const MindmapNode({
    required this.id,
    this.nodeId,
    this.workspaceId,
    this.parentId,
    required this.text,
    required this.position,
    required this.color,
    this.width = 120,
    this.height = 60,
    this.level = 0,
    this.contentType = NodeContentType.text,
    this.contentUrl,
    this.description,
    this.analysisStatus,
    this.createdAt,
    this.updatedAt,
  });

  /// 노드의 중심점을 반환합니다
  /// position이 이미 노드의 중심 좌표입니다
  Offset get center => position;

  MindmapNode copyWith({
    String? id,
    int? nodeId,
    int? workspaceId,
    int? parentId,
    String? text,
    Offset? position,
    Color? color,
    double? width,
    double? height,
    int? level,
    NodeContentType? contentType,
    String? contentUrl,
    String? description,
    String? analysisStatus,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return MindmapNode(
      id: id ?? this.id,
      nodeId: nodeId ?? this.nodeId,
      workspaceId: workspaceId ?? this.workspaceId,
      parentId: parentId ?? this.parentId,
      text: text ?? this.text,
      position: position ?? this.position,
      color: color ?? this.color,
      width: width ?? this.width,
      height: height ?? this.height,
      level: level ?? this.level,
      contentType: contentType ?? this.contentType,
      contentUrl: contentUrl ?? this.contentUrl,
      description: description ?? this.description,
      analysisStatus: analysisStatus ?? this.analysisStatus,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: backend/mindmap-service/src/main/java/com/ssafy/mindmapservice/domain/MindmapNode.java
package com.ssafy.mindmapservice.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;

/**
 * 마인드맵 노드 엔티티
 * MongoDB 컬렉션: nodes
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "nodes")
public class MindmapNode {

    /**
     * MongoDB 내부 ID
     */
    @Id
    private String id;

    /**
     * 노드 고유 ID (서버에서 자동 생성)
     * 워크스페이스별로 1부터 자동 증가
     */
    @Indexed
    private Long nodeId;

    /**
     * 워크스페이스 ID (파티셔닝 키)
     */
    @Indexed
    private Long workspaceId;

    /**
     * 부모 노드 ID (트리 구조)
     * 루트 노드는 null
     */
    private Long parentId;

    /**
     * 노드 타입
     * 예: "text", "image", "video"
     */
    private String type;

    /**
     * 노드 키워드/제목
     */
    private String keyword;

    /**
     * 노드 메모 내용
     */
    private String memo;

    /**
     * AI 분석 상태
     * NONE: 분석 불필요 (텍스트 노드)
     * PENDING: 분석 요청됨
     * PROCESSING: 분석 중
     * DONE: 분석 완료
     * FAILED: 분석 실패
     */
    private AnalysisStatus analysisStatus;

    /**
     * 캔버스 X 좌표
     */
    private Double x;

    /**
     * 캔버스 Y 좌표
     */
    private Double y;

    /**
     * 노드 색상 (hex code)
     * 예: "#ffd700", "#3b82f6"
     */
    private String color;

    /**
     * 생성 시각
     */
    @CreatedDate
    private LocalDateTime createdAt;

    /**
     * 수정 시각
     */
    @LastModifiedDate
    private LocalDateTime updatedAt;

    /**
     * AI 분석 상태 Enum
     */
    public enum AnalysisStatus {
        NONE,       // 분석 불필요
        PENDING,    // 분석 대기
        PROCESSING, // 분석 중
        DONE,       // 분석 완료
        FAILED      // 분석 실패
    }
}
